src/julia.h

```diff
@@ -73,13 +73,18 @@ extern "C" {
 #ifndef JULIA_ENABLE_THREADING
 // Definition for compiling non-thread-safe Julia.
 #  define JL_THREAD
-#elif !defined(_OS_WINDOWS_)
+#elif !defined(_COMPILER_MICROSOFT_)
 // Definition for compiling Julia on platforms with GCC __thread.
 #  define JL_THREAD __thread
 #else
 // Definition for compiling Julia on Windows
 #  define JL_THREAD __declspec(thread)
 #endif
+#ifdef _COMPILER_MICROSOFT_
+#define JL_THREAD_EXPORT JL_THREAD
+#else
+#define JL_THREAD_EXPORT DLLEXPORT JL_THREAD
+#endif
```

Comment:
rebased on top of the other related pr's and this is now the most substantive remaining part of the pr, though since it's still nonfunctional on windows it can probably wait until a TLS refactoring.